# AI-assisted SQL for Absolute Beginners

Welcome to today's webinar! We are going to work with AI to write SQL examining data covering London Fire Brigade's animal rescue efforts.

<br>

<center><img src="fire_fighter_chatgpt.png" width=300></center>

<br>

### 📊 Our data

This project will use a simplified version of the data from the [London Fire Brigade](https://data.london.gov.uk/dataset/animal-rescue-incidents-attended-by-lfb-2ogkn/). The data is attached to this workbook in the form of a csv.

<br>

**Table: `animal_rescue_incidents.csv`**
| Column | Description |
|---|---|
| `incident_id` | Unique identifier for the incident. |
| `datetime` | Date and time of the rescue. |
| `animal_type` | Species of the animal that was rescued. |
| `location_type` | Environment where the rescue occurred. |
| `rescue_type` | Type of assistance provided. |
| `area_code` | Geographic code. |
| `area` | Name of the London borough. |
| `nb_units_deployed` | Number of fire units/vehicles used. |
| `service_duration_hrs` | Total time of the operation. |
| `hourly_cost` | Cost per hour for the rescue. |

## Goals of this Webinar
In this beginner webinar we will learn what SQL is, how to generate it with AI assistance via prompts, and how to check the AI's output to ensure quality. We will cover data inspection, filtering, aggregation, ranking and sorting, as well as a number of tips and tricks for getting the most out of your AI assistant.

### 🤖 The AI Workflow
When solving a question, follow the following steps:
1. Think about the data you have and what you want to achieve.
2. Write a natural language prompt.
3. Understand the code generated by the AI.
4. Check the AI’s code for errors or hallucinations.
5. Run the code and verify the results.
6. Iterate.

## 1. What is SQL?

SQL stands for **Structured Query Language**. It's a programming language used to fetch and transform data.

Underneath this text is a SQL cell - it contains SQL code, and executes it upon demand. You run it by clicking 'Run' on the right hand side of the cell. Don't worry about the contents of this cell, just run it.
### Press Run on the SQL Cell below

In [1]:
SELECT
	animal_type,
	COUNT(*) AS incident_count
FROM 'animal_rescue_incidents.csv'
WHERE location_type = 'Outdoor'
GROUP BY animal_type
ORDER BY incident_count DESC
LIMIT 5;

animal_type incident_count
1 cat         1341          
2 dog          690          
3 bird         503          
4 fox          216          
5 horse        181

SQL is used for querying data sets, so the input and output are always data sets. They have a specific structure:
* The columns of the data set are called 'Fields'. `animal_type` and `incident_count` in the above output data set are `fields`.
* The rows of the data set are sometimes called 'Observations'. DataLab puts an index next to each observation, labeling them 1 to 5 in this case.

## 2. Data Inspection
Generative AI is very good at converting natural language into SQL statements. There's nothing stopping you from writing SQL directly, but AI is a powerful tool, especially if you're not sure where to start. It can help you write complex SQL from a brief description of the desired outcome.
<br>
<br>The first step of any SQL query is to look at the data we have available. Our table is called `animal_rescue_incidents.csv`.
### Ask the AI to reveal some rows of this table.

Remember to follow the AI Workflow 🤖:
1. Think about the data you have and what you want to achieve.
2. Write a natural language prompt.
3. Understand the code generated by the AI.
4. Check the AI’s code for errors or hallucinations.
5. Run the code and verify the results.
6. Iterate.

In [2]:
SELECT * FROM information_schema.tables;

table_catalog table_schema table_name table_type self_referencing_column_name
1 temp          main         q1_example VIEW       NA                          
  reference_generation user_defined_type_catalog user_defined_type_schema
1 NA                   NA                        NA                      
  user_defined_type_name is_insertable_into is_typed commit_action
1 NA                     NO                 NO       NA           
  TABLE_COMMENT
1 NA

Answer: <br>
Show me some rows of 'animal_rescue_incidents.csv'

In [6]:
SELECT *
FROM 'animal_rescue_incidents.csv'
LIMIT 10;

incident_id datetime         animal_type                     
1  139091      01/01/2009 03:01 dog                             
2  275091      01/01/2009 08:51 fox                             
3  2075091     04/01/2009 10:07 dog                             
4  2872091     05/01/2009 12:27 horse                           
5  3553091     06/01/2009 15:23 rabbit                          
6  3742091     06/01/2009 19:30 unknown - heavy livestock animal
7  4011091     07/01/2009 06:29 dog                             
8  4211091     07/01/2009 11:55 dog                             
9  4306091     07/01/2009 13:48 squirrel                        
10 4715091     07/01/2009 21:24 dog                             
   location_type     rescue_type                     area_code
1  Dwelling          Other animal assistance         E09000008
2  Outdoor Structure Other animal assistance         E09000008
3  Outdoor Structure Animal rescue from below ground E09000029
4  Non Residential   Animal rescue from water        E09000017
5  Dwelling          Other animal assistance         E09000016
6  Dwelling          Other animal assistance         E09000002
7  Outdoor           Other animal assistance         E09000031
8  Outdoor           Animal rescue from water        E09000026
9  Dwelling          Animal rescue from height       E09000008
10 Outdoor           Animal rescue from water        E09000012
   area                 incident_description nb_units_deployed
1  Croydon              Redacted             1                
2  Croydon              Redacted             1                
3  Sutton               Redacted             1                
4  Hillingdon           Redacted             1                
5  Havering             Redacted             1                
6  Barking And Dagenham Redacted             1                
7  Waltham Forest       Redacted             1                
8  Redbridge            Redacted             1                
9  Croydon              Redacted             1                
10 Hackney              Redacted             1                
   service_duration_hrs hourly_cost
1  2                    255        
2  1                    255        
3  1                    255        
4  1                    255        
5  1                    255        
6  1                    255        
7  1                    255        
8  1                    255        
9  1                    255        
10 1                    255

The Keywords the AI should have used are `SELECT` and `FROM`.
<br>`SELECT` is the first word of most SQL statements. It ins an instruction to fetch data, and is followed by a list of what data we want, or by `*` if we want to get everything in a given table.
<br> `FROM` is used before the name of the source table that we want to read.
<br>
<br>![select_from_image](select_from.png)
<br>
<br> The AI may also choose to use `LIMIT`, which specifies a number of rows to return. Requesting less data can make a query much faster.
<br>It may also have chosen to put a semicolon (`;`) at the end. Don't worry about this, it's only important if you are writing multiple SQL statements at once and need to show where one of them ends.
<br>
A prompt like 'show me the data' is unlikely to work, because it doesn't tell the AI which data we're referring to. It's best if we refer to the table we want by name.

> 🔍 Troubleshooting: Make sure that the AI uses the correct name of the table (animal_rescue_incidents.csv).
> <br>The AI might try to look at `animal_rescue_incidents` instead of `animal_rescue_incidents.csv`. When this happens, we can adjust our prompt, or we can edit the SQL code directly.

We don't always want every field from the data.
### Ask the AI to fetch just the animal_type and datetime fields.

Answer: <br>
Show me animal_type and datetime from 'animal_rescue_incidents.csv'

In [5]:
SELECT animal_type, datetime
FROM 'animal_rescue_incidents.csv';

animal_type                      datetime        
1     dog                              01/01/2009 03:01
2     fox                              01/01/2009 08:51
3     dog                              04/01/2009 10:07
4     horse                            05/01/2009 12:27
5     rabbit                           06/01/2009 15:23
6     unknown - heavy livestock animal 06/01/2009 19:30
7     dog                              07/01/2009 06:29
8     dog                              07/01/2009 11:55
9     squirrel                         07/01/2009 13:48
10    dog                              07/01/2009 21:24
11    cat                              08/01/2009 14:34
12    bird                             08/01/2009 19:23
13    dog                              09/01/2009 11:01
14    dog                              09/01/2009 11:18
15    cat                              09/01/2009 12:40
16    cat                              09/01/2009 13:43
17    cat                              09/01/2009 14:30
18    dog                              09/01/2009 14:39
19    NA                               10/01/2009 09:35
20    dog                              10/01/2009 10:09
21    cat                              10/01/2009 11:27
22    cat                              10/01/2009 12:26
23    dog                              10/01/2009 12:32
24    cat                              10/01/2009 13:05
25    dog                              10/01/2009 13:28
26    cat                              10/01/2009 16:38
27    dog                              11/01/2009 10:11
28    bird                             11/01/2009 12:32
29    cat                              12/01/2009 13:10
30    dog                              12/01/2009 16:39
⋮     ⋮                                ⋮               
13043 bird                             20/12/2025 16:31
13044 cat                              20/12/2025 18:27
13045 fox                              20/12/2025 23:51
13046 cat                              21/12/2025 14:05
13047 dog                              21/12/2025 14:49
13048 dog                              21/12/2025 22:37
13049 pigeon                           22/12/2025 13:21
13050 bird                             22/12/2025 14:11
13051 snake                            22/12/2025 22:59
13052 cat                              23/12/2025 14:34
13053 cat                              23/12/2025 18:11
13054 cat                              23/12/2025 20:19
13055 bird                             24/12/2025 14:49
13056 cat                              24/12/2025 16:28
13057 cat                              25/12/2025 02:27
13058 cat                              25/12/2025 16:39
13059 cat                              25/12/2025 18:07
13060 cat                              25/12/2025 22:34
13061 cat                              27/12/2025 10:45
13062 NA                               27/12/2025 12:04
13063 cat                              27/12/2025 19:26
13064 cat                              27/12/2025 21:36
13065 cat                              28/12/2025 12:05
13066 bird                             28/12/2025 12:44
13067 fox                              29/12/2025 09:39
13068 cat                              29/12/2025 14:07
13069 dog                              29/12/2025 20:26
13070 dog                              30/12/2025 14:08
13071 cat                              30/12/2025 20:59
13072 cat                              31/12/2025 05:28

Note that when fetching specific fields in the `SELECT` clause, they are separated by a comma. The AI might have arranged the fields on different lines, or kept them on the same line as `SELECT`. Both of these are fine. Line breaks don't have any meaning in SQL, they are just used to make code easier to read.

> 🔍 Troubleshooting: As well as the table name, check the field names against the data we just pulled, or the schema above.

## 3. Filtering Data
Now that we know what the data looks like, we can use a filter to focus on the parts of the data that we are most interested in.

### Ask the AI to show you the incidents that were about cats

Answer: <br>
Show me the incidents in 'animal_rescue_incidents.csv' where animal_type is 'cat'

In [7]:
SELECT *
FROM 'animal_rescue_incidents.csv'
WHERE animal_type = 'cat';

incident_id     datetime         animal_type location_type    
1    5186091         08/01/2009 14:34 cat         Outdoor          
2    5724091         09/01/2009 12:40 cat         Outdoor          
3    5770091         09/01/2009 13:43 cat         Outdoor          
4    5789091         09/01/2009 14:30 cat         Outdoor          
5    6317091         10/01/2009 11:27 cat         Outdoor          
6    6353091         10/01/2009 12:26 cat         Dwelling         
7    6378091         10/01/2009 13:05 cat         Non Residential  
8    6530091         10/01/2009 16:38 cat         Outdoor          
9    7613091         12/01/2009 13:10 cat         Dwelling         
10   10829091        18/01/2009 11:08 cat         Dwelling         
11   10976091        18/01/2009 16:19 cat         Outdoor          
12   14002091        24/01/2009 10:31 cat         Dwelling         
13   14652091        25/01/2009 10:26 cat         Outdoor          
14   15682091        27/01/2009 11:21 cat         Dwelling         
15   16245091        28/01/2009 13:40 cat         Outdoor          
16   16271091        28/01/2009 14:41 cat         Dwelling         
17   18369091        01/02/2009 08:48 cat         Dwelling         
18   20379091        04/02/2009 17:47 cat         Dwelling         
19   21319091        06/02/2009 14:42 cat         Outdoor          
20   21401091        06/02/2009 17:14 cat         Dwelling         
21   21458091        06/02/2009 19:29 cat         Dwelling         
22   22115091        08/02/2009 00:59 cat         Other Residential
23   27805091        17/02/2009 13:23 cat         Outdoor          
24   29403091        20/02/2009 09:35 cat         Dwelling         
25   30194091        21/02/2009 16:12 cat         Outdoor          
26   169092          23/02/2009 15:28 cat         Dwelling         
27   32555091        25/02/2009 22:18 cat         Dwelling         
28   35246091        02/03/2009 14:29 cat         Dwelling         
29   35738091        03/03/2009 11:05 cat         Dwelling         
30   37589091        06/03/2009 15:22 cat         Outdoor          
⋮    ⋮               ⋮                ⋮           ⋮                
6855 234707-16122025 16/12/2025 09:11 cat         Road Vehicle     
6856 234733-16122025 16/12/2025 10:04 cat         Dwelling         
6857 234789-16122025 16/12/2025 11:34 cat         Dwelling         
6858 234818-16122025 16/12/2025 12:19 cat         Outdoor Structure
6859 234953-16122025 16/12/2025 16:18 cat         Road Vehicle     
6860 235589-17122025 17/12/2025 17:30 cat         Dwelling         
6861 236280-18122025 18/12/2025 21:40 cat         Dwelling         
6862 236318-18122025 18/12/2025 23:02 cat         Dwelling         
6863 236510-19122025 19/12/2025 10:26 cat         Outdoor          
6864 236643-19122025 19/12/2025 14:57 cat         Dwelling         
6865 237144-20122025 20/12/2025 10:11 cat         Outdoor          
6866 237237-20122025 20/12/2025 13:18 cat         Dwelling         
6867 237335-20122025 20/12/2025 16:07 cat         Dwelling         
6868 237415-20122025 20/12/2025 18:27 cat         Dwelling         
6869 237846-21122025 21/12/2025 14:05 cat         Dwelling         
6870 239101-23122025 23/12/2025 14:34 cat         Dwelling         
6871 239244-23122025 23/12/2025 18:11 cat         Outdoor Structure
6872 239359-23122025 23/12/2025 20:19 cat         Dwelling         
6873 239893-24122025 24/12/2025 16:28 cat         Dwelling         
6874 240198-25122025 25/12/2025 02:27 cat         Dwelling         
6875 240593-25122025 25/12/2025 16:39 cat         Dwelling         
6876 240656-25122025 25/12/2025 18:07 cat         Dwelling         
6877 240791-25122025 25/12/2025 22:34 cat         Dwelling         
6878 241608-27122025 27/12/2025 10:45 cat         Dwelling         
6879 241863-27122025 27/12/2025 19:26 cat         Dwelling         
6880 241937-27122025 27/12/2025 21:36 cat         Outdoor          
6881 242211-28122025 28/12/2025 12:05 cat         Outdoor    

The AI should have used a `WHERE` clause. `Where` is used alonside a true/false statement to filter the data to cases where that statement is true.

![where image](where.png)

> 🔍 Troubleshooting: Check that the casing of 'cat' (as opposed to 'Cat') matches the way it is presented in data.

### Ask the AI to show you the incidents that weren't about cats

Answer: <br>
Show me the incidents in 'animal_rescue_incidents.csv' where animal_type is not 'cat'

In [8]:
SELECT *
FROM 'animal_rescue_incidents.csv'
WHERE animal_type <> 'cat';

incident_id     datetime         animal_type                     
1    139091          01/01/2009 03:01 dog                             
2    275091          01/01/2009 08:51 fox                             
3    2075091         04/01/2009 10:07 dog                             
4    2872091         05/01/2009 12:27 horse                           
5    3553091         06/01/2009 15:23 rabbit                          
6    3742091         06/01/2009 19:30 unknown - heavy livestock animal
7    4011091         07/01/2009 06:29 dog                             
8    4211091         07/01/2009 11:55 dog                             
9    4306091         07/01/2009 13:48 squirrel                        
10   4715091         07/01/2009 21:24 dog                             
11   5363091         08/01/2009 19:23 bird                            
12   5682091         09/01/2009 11:01 dog                             
13   5688091         09/01/2009 11:18 dog                             
14   5797091         09/01/2009 14:39 dog                             
15   6270091         10/01/2009 10:09 dog                             
16   6355091         10/01/2009 12:32 dog                             
17   6400091         10/01/2009 13:28 dog                             
18   6970091         11/01/2009 10:11 dog                             
19   7051091         11/01/2009 12:32 bird                            
20   7720091         12/01/2009 16:39 dog                             
21   8648091         14/01/2009 06:38 dog                             
22   9252091         15/01/2009 12:38 bird                            
23   10299091        17/01/2009 11:40 dog                             
24   10470091        17/01/2009 18:08 dog                             
25   12317091        21/01/2009 00:43 dog                             
26   12451091        21/01/2009 10:33 fox                             
27   12533091        21/01/2009 14:04 dog                             
28   12770091        21/01/2009 21:55 fox                             
29   13245091        22/01/2009 21:35 bird                            
30   14691091        25/01/2009 12:11 dog                             
⋮    ⋮               ⋮                ⋮                               
5894 223509-28112025 28/11/2025 14:18 bird                            
5895 224757-30112025 30/11/2025 15:15 bird                            
5896 225263-01122025 01/12/2025 13:25 fox                             
5897 225911-02122025 02/12/2025 14:39 dog                             
5898 226172-02122025 02/12/2025 22:13 dog                             
5899 226759-03122025 03/12/2025 19:29 dog                             
5900 226833-03122025 03/12/2025 20:58 bird                            
5901 226966-04122025 04/12/2025 04:30 deer                            
5902 229203-07122025 07/12/2025 14:54 bird                            
5903 229897-08122025 08/12/2025 16:11 dog                             
5904 230429-09122025 09/12/2025 12:56 fox                             
5905 230535-09122025 09/12/2025 16:27 bird                            
5906 231078-10122025 10/12/2025 13:17 fox                             
5907 231812-11122025 11/12/2025 18:26 bird                            
5908 232328-12122025 12/12/2025 15:29 bird                            
5909 233762-14122025 14/12/2025 19:21 bird                            
5910 234133-15122025 15/12/2025 12:49 bird                            
5911 236689-19122025 19/12/2025 16:04 bird                            
5912 237345-20122025 20/12/2025 16:31 bird                            
5913 237585-20122025 20/12/2025 23:51 fox                             
5914 237873-21122025 21/12/2025 14:49 dog                             
5915 238105-21122025 21/12/2025 22:37 dog                             
5916 238389-22122025 22/12/2025 13:21 pigeon                          
5917 238422-22122025 22/12/2025 14:11 bird                            
5918 238766-22122025 22/12/20

Out of 13,072 total rows there are 6,884 `WHERE animal_type = 'cat'` and 5,923 `WHERE animal_type != 'cat'`. Not everything in this data is either a Cat or not a Cat!
### Ask the AI to show the incidents where we don't know the animal_type.

Answer: <br>
Show me the incidents in 'animal_rescue_incidents.csv' where we don't know the animal_type

In [9]:
SELECT *
FROM 'animal_rescue_incidents.csv'
WHERE animal_type IS NULL OR animal_type = '';

incident_id     datetime         animal_type location_type    
1   6259091         10/01/2009 09:35 NA          Outdoor          
2   14784091        25/01/2009 15:27 NA          Outdoor          
3   30981091        22/02/2009 17:22 NA          Dwelling         
4   60105091        11/04/2009 11:47 NA          Dwelling         
5   92294091        31/05/2009 14:28 NA          Dwelling         
6   107370091       22/06/2009 13:43 NA          Dwelling         
7   163350091       08/09/2009 22:08 NA          Outdoor          
8   201050091       05/11/2009 15:38 NA          Dwelling         
9   224156091       15/12/2009 10:49 NA          Outdoor          
10  231733091       28/12/2009 11:52 NA          Dwelling         
11  29349101        24/02/2010 15:36 NA          Dwelling         
12  33402101        04/03/2010 10:07 NA          Outdoor          
13  24109           06/03/2010 14:58 NA          Outdoor          
14  49100101        02/04/2010 12:56 NA          Outdoor          
15  65274101        29/04/2010 10:40 NA          Dwelling         
16  77957101        20/05/2010 21:35 NA          Dwelling         
17  84569101        30/05/2010 19:44 NA          Outdoor          
18  85659101        01/06/2010 18:54 NA          Dwelling         
19  95038101        17/06/2010 17:11 NA          Dwelling         
20  95570101        18/06/2010 15:26 NA          Outdoor          
21  96413101        19/06/2010 21:50 NA          Outdoor          
22  103128101       29/06/2010 09:25 NA          Dwelling         
23  103894101       30/06/2010 11:29 NA          Outdoor Structure
24  110956101       08/07/2010 18:39 NA          Dwelling         
25  124199101       24/07/2010 21:07 NA          Dwelling         
26  137703101       12/08/2010 08:31 NA          Dwelling         
27  151656101       04/09/2010 17:14 NA          Outdoor          
28  162969101       24/09/2010 11:14 NA          Dwelling         
29  1790111         04/01/2011 17:30 NA          Dwelling         
30  1842111         04/01/2011 18:56 NA          Outdoor          
⋮   ⋮               ⋮                ⋮           ⋮                
236 141602-13092023 13/09/2023 22:02 NA          Dwelling         
237 144309-18092023 18/09/2023 09:02 NA          Dwelling         
238 164161-20102023 20/10/2023 22:00 NA          Dwelling         
239 165526-23102023 23/10/2023 10:07 NA          Dwelling         
240 168295-28102023 28/10/2023 10:42 NA          Dwelling         
241 194658-11122023 11/12/2023 14:20 NA          Dwelling         
242 199945-20122023 20/12/2023 16:31 NA          Dwelling         
243 002673-04012024 04/01/2024 18:59 NA          Dwelling         
244 018890-01022024 01/02/2024 13:52 NA          Outdoor          
245 019995-03022024 03/02/2024 12:25 NA          Dwelling         
246 020359-04022024 04/02/2024 00:37 NA          Dwelling         
247 021373-05022024 05/02/2024 21:10 NA          Dwelling         
248 024033-10022024 10/02/2024 16:46 NA          Outdoor          
249 024719-11022024 11/02/2024 23:15 NA          Dwelling         
250 042286-15032024 15/03/2024 12:37 NA          Dwelling         
251 071543-06052024 06/05/2024 09:55 NA          Dwelling         
252 142323-25082024 25/08/2024 18:31 NA          Outdoor Structure
253 167724-03102024 03/10/2024 15:39 NA          Outdoor          
254 215329-19122024 19/12/2024 23:25 NA          Outdoor          
255 065858-17042025 17/04/2025 15:52 NA          Road Vehicle     
256 091591-25052025 25/05/2025 07:01 NA          Dwelling         
257 096993-02062025 02/06/2025 15:23 NA          Dwelling         
258 126629-10072025 10/07/2025 15:38 NA          Outdoor Structure
259 139470-27072025 27/07/2025 00:18 NA          Dwelling         
260 147490-08082025 08/08/2025 14:28 NA          Dwelling         
261 158263-22082025 22/08/2025 20:07 NA          Dwelling         
262 174660-16092025 16/09/2025 16:04 NA          Road Vehicle     
263 194345-15102025 15/10/2025 19:51 NA          Ro

There are some entries where the animal type is `NULL`. This isn't text. It's a special type of unknown value that doesn't equal anything. That means it won't show up when you filter for Cats, or exclude Cats, unless you mention it specifically.
<br>![null image](null_table.png)

> 🔍 Troubleshooting: Make sure that the code is specifically checking `animal_type IS NULL`, rather than looking for equality such as `animal_type = 'null'` or `animal_type = 'unknown'`.

### Ask the AI to show you incidents where it wasn't a cat, _or_ we don't know what it was.

Answer: <br>
Show me the incidents in 'animal_rescue_incidents.csv' where animal_type wasn't 'cat' or we don't know the animal_type

In [10]:
SELECT *
FROM 'animal_rescue_incidents.csv'
WHERE animal_type IS NULL OR animal_type <> 'cat';

incident_id     datetime         animal_type location_type    
1    6259091         10/01/2009 09:35 NA          Outdoor          
2    14784091        25/01/2009 15:27 NA          Outdoor          
3    30981091        22/02/2009 17:22 NA          Dwelling         
4    60105091        11/04/2009 11:47 NA          Dwelling         
5    92294091        31/05/2009 14:28 NA          Dwelling         
6    107370091       22/06/2009 13:43 NA          Dwelling         
7    163350091       08/09/2009 22:08 NA          Outdoor          
8    201050091       05/11/2009 15:38 NA          Dwelling         
9    224156091       15/12/2009 10:49 NA          Outdoor          
10   231733091       28/12/2009 11:52 NA          Dwelling         
11   29349101        24/02/2010 15:36 NA          Dwelling         
12   33402101        04/03/2010 10:07 NA          Outdoor          
13   24109           06/03/2010 14:58 NA          Outdoor          
14   49100101        02/04/2010 12:56 NA          Outdoor          
15   65274101        29/04/2010 10:40 NA          Dwelling         
16   77957101        20/05/2010 21:35 NA          Dwelling         
17   84569101        30/05/2010 19:44 NA          Outdoor          
18   85659101        01/06/2010 18:54 NA          Dwelling         
19   95038101        17/06/2010 17:11 NA          Dwelling         
20   95570101        18/06/2010 15:26 NA          Outdoor          
21   96413101        19/06/2010 21:50 NA          Outdoor          
22   103128101       29/06/2010 09:25 NA          Dwelling         
23   103894101       30/06/2010 11:29 NA          Outdoor Structure
24   110956101       08/07/2010 18:39 NA          Dwelling         
25   124199101       24/07/2010 21:07 NA          Dwelling         
26   137703101       12/08/2010 08:31 NA          Dwelling         
27   151656101       04/09/2010 17:14 NA          Outdoor          
28   162969101       24/09/2010 11:14 NA          Dwelling         
29   1790111         04/01/2011 17:30 NA          Dwelling         
30   1842111         04/01/2011 18:56 NA          Outdoor          
⋮    ⋮               ⋮                ⋮           ⋮                
6159 223509-28112025 28/11/2025 14:18 bird        Outdoor          
6160 224757-30112025 30/11/2025 15:15 bird        Non Residential  
6161 225263-01122025 01/12/2025 13:25 fox         Outdoor          
6162 225911-02122025 02/12/2025 14:39 dog         Dwelling         
6163 226172-02122025 02/12/2025 22:13 dog         Dwelling         
6164 226759-03122025 03/12/2025 19:29 dog         Outdoor Structure
6165 226833-03122025 03/12/2025 20:58 bird        Non Residential  
6166 226966-04122025 04/12/2025 04:30 deer        Outdoor          
6167 229203-07122025 07/12/2025 14:54 bird        Outdoor          
6168 229897-08122025 08/12/2025 16:11 dog         Dwelling         
6169 230429-09122025 09/12/2025 12:56 fox         Dwelling         
6170 230535-09122025 09/12/2025 16:27 bird        Outdoor          
6171 231078-10122025 10/12/2025 13:17 fox         Non Residential  
6172 231812-11122025 11/12/2025 18:26 bird        Dwelling         
6173 232328-12122025 12/12/2025 15:29 bird        Dwelling         
6174 233762-14122025 14/12/2025 19:21 bird        Non Residential  
6175 234133-15122025 15/12/2025 12:49 bird        Non Residential  
6176 236689-19122025 19/12/2025 16:04 bird        Outdoor Structure
6177 237345-20122025 20/12/2025 16:31 bird        Dwelling         
6178 237585-20122025 20/12/2025 23:51 fox         Outdoor          
6179 237873-21122025 21/12/2025 14:49 dog         Dwelling         
6180 238105-21122025 21/12/2025 22:37 dog         Outdoor          
6181 238389-22122025 22/12/2025 13:21 pigeon      Outdoor Structure
6182 238422-22122025 22/12/2025 14:11 bird        Outdoor Structure
6183 238766-22122025 22/12/2025 22:59 snake       Non Residential  
6184 239828-24122025 24/12/2025 14:49 bird        Dwelling         
6185 242223-28122025 28/12/2025 12:44 bird        Outdoor    

 `OR` and `AND` are used to combine truth conditions together.
 <br>`Condition A OR Condition B` is true if either of the Conditions are true.
  
![or operator](or_operator.png)
<br>
<br>`Condition A AND Condition B` is true if both of the Conditions are true.
![and operator](and_operator.png)

> 🔍 Troubleshooting: Once again, make sure that the code is specifically checking `animal_type IS NULL`, rather than looking for equality such as `animal_type = 'null'` or `animal_type = 'unknown'`.

## 4. Aggregation

So far we've only looked at the data. It's time to start calculating.

### Ask the AI to tell you how many rows there are in the data.

Answer: <br>
How many rows are there in 'animal_rescue_incidents.csv'?

In [11]:
SELECT COUNT(*) FROM 'animal_rescue_incidents.csv';

count_star()
1 13072

`COUNT(*)` simply tells us the number of rows in the data.

### Ask the AI to tell you how many incidents involved a dog.

Answer: <br>
How many rows are there in 'animal_rescue_incidents.csv' where animal_type is 'dog'?

In [12]:
SELECT COUNT(*) AS dog_incident_count
FROM 'animal_rescue_incidents.csv'
WHERE animal_type = 'dog';

dog_incident_count
1 1756

The AI may have named the count, perhaps `dog_incident_count`. This is called 'Aliasing', and it's when we specify the name (or 'alias) that a field will take in the output data set.
<br>
<br>
It's helpful to know how many incidents involved a dog, but what if we want a full breakdown of incidents by animal type? Would we have to run a separate piece of code for each one? What if we don't know all the types? There's a simpler solution.
### Ask the AI to get you a count of incidents by animal_type.

Answer: <br>
Get me a count of rows in 'animal_rescue_incidents.csv' by animal_type

In [13]:
SELECT animal_type, COUNT(*) AS incident_count
FROM 'animal_rescue_incidents.csv'
GROUP BY animal_type;

animal_type                      incident_count
1  horse                             233          
2  deer                              223          
3  lizard                              6          
4  fish                                3          
5  budgie                              4          
6  ferret                             12          
7  rat                                 2          
8  fox                               762          
9  unknown - heavy livestock animal   51          
10 squirrel                          105          
11 cow                                 9          
12 cat                              6884          
13 unknown - wild animal             135          
14 hedgehog                            6          
15 pigeon                              6          
16 goat                                4          
17 sheep                               7          
18 bull                                1          
19 rabbit                             21          
20 bird                             2531          
21 NA                                265          
22 dog                              1756          
23 snake                              22          
24 hamster                            20          
25 lamb                                2          
26 tortoise                            2

This is an aggregation. The `GROUP BY animal_type` clause establishes that we want to separate the table into each `animal_type`, and then we run `COUNT(*)` over each segment.

> 🔍 Troubleshooting: Make sure that the field is correctly referred to as `animal_type` in both the `SELECT` and `GROUP BY` clauses.

We can aggregate by more than one field at once.
### Ask the AI to count combinations of `location_type` and `rescue_type`.

Answer: <br>
Get me a count of rows in 'animal_rescue_incidents.csv' by location_type and rescue_type

In [18]:
SELECT
  location_type,
  rescue_type,
  COUNT(*) AS incident_count
FROM
  'animal_rescue_incidents.csv'
GROUP BY
  location_type,
  rescue_type;

location_type     rescue_type                     incident_count
1  Outdoor           Animal rescue from water         481          
2  Dwelling          Animal rescue from height       2987          
3  Non Residential   Other animal assistance          606          
4  Road Vehicle      Animal rescue from below ground   10          
5  Other Residential Animal rescue from below ground    5          
6  Dwelling          Other animal assistance         3302          
7  Outdoor Structure Animal rescue from below ground   84          
8  Non Residential   Animal rescue from height        580          
9  Outdoor Structure Other animal assistance          525          
10 Dwelling          Animal rescue from below ground  742          
11 Road Vehicle      Animal rescue from height          7          
12 Other Residential Animal rescue from height         18          
13 Non Residential   Animal rescue from water           5          
14 Outdoor           Other animal assistance         1154          
15 Road Vehicle      Other animal assistance          513          
16 Non Residential   Animal rescue from below ground  115          
17 Other Residential Other animal assistance           19          
18 Outdoor Structure Animal rescue from height        299          
19 Boat              Animal rescue from water           2          
20 Outdoor           Animal rescue from height       1304          
21 Dwelling          Animal rescue from water          36          
22 Outdoor           Animal rescue from below ground  259          
23 Road Vehicle      Animal rescue from water           3          
24 Boat              Other animal assistance            2          
25 Outdoor Structure Animal rescue from water          14

That should show us every combination of `location_type` and `rescue_type` in the table, and a count of how many times it ocurrs. When checking the AI's SQL, notice that just like the the `SELECT` clause, the elements of the `GROUP BY` clause should be separated by commas.

> 🔍 Troubleshooting: Make sure that the AI has used the correct field names, `location_type` and `rescue_type`, in both the `SELECT` and `GROUP BY` clauses.

## 5. Ranking & Sorting

Did you notice that the data was in random order in the last output? Or did the AI sort it for you? SQL code won't sort the output unless it was specified to do so.
### Ask the AI to show you a count of rows by animal_type sorted by the number of incidents.

Answer: <br>
Show me a count of rows in 'animal_rescue_incidents.csv' by animal_type ordered by the number of incidents

In [20]:
SELECT animal_type, COUNT(*) AS incident_count
FROM 'animal_rescue_incidents.csv'
GROUP BY animal_type
ORDER BY incident_count DESC;

animal_type                      incident_count
1  cat                              6884          
2  bird                             2531          
3  dog                              1756          
4  fox                               762          
5  NA                                265          
6  horse                             233          
7  deer                              223          
8  unknown - wild animal             135          
9  squirrel                          105          
10 unknown - heavy livestock animal   51          
11 snake                              22          
12 rabbit                             21          
13 hamster                            20          
14 ferret                             12          
15 cow                                 9          
16 sheep                               7          
17 lizard                              6          
18 hedgehog                            6          
19 pigeon                              6          
20 goat                                4          
21 budgie                              4          
22 fish                                3          
23 rat                                 2          
24 lamb                                2          
25 tortoise                            2          
26 bull                                1

The `ORDER BY` clause tells SQL to sort, or order, the output. The AI may or may not have added `DESC`. Without it, the data will be sorted in ascending order - ie. 1, 2, 3 and so on. Adding `DESC` will instead sort the data in descending order, so 3, 2, 1.
<br>
<br>Let's go one step further
### Ask the AI to show you the top 5 Animal Types by number of instances

Answer: <br>
Show me a the top 5 animal_type entries in 'animal_rescue_incidents.csv' by number of incidents

In [21]:
SELECT animal_type, COUNT(*) AS incident_count
FROM 'animal_rescue_incidents.csv'
GROUP BY animal_type
ORDER BY incident_count DESC
LIMIT 5;

animal_type incident_count
1 cat         6884          
2 bird        2531          
3 dog         1756          
4 fox          762          
5 NA           265

The AI should have used:
* `SELECT` and `FROM` to set up the query
* `COUNT(*)` and `GROUP BY` to aggregate the incidents
* `ORDER BY`, `DESC` and `LIMIT` to fetch only the top entries

> 🔍 Troubleshooting: As always, check that the AI has used the correct names for the table and fields.

<br>This is a complex SQL query, but it's easy to conceptualise and ask the AI to write for you.

## 6. Other Forms of AI Assistance - Fixing and Explaining
Another very helpful way to use AI when writing SQL is to get it to fix statements for you. Try running the following code. Does the error message make any sense to you?

In [ ]:
SELECT animal_type COUNT(*) AS incident_count
FROM 'animal_rescue_incidents.csv'
ORDER BY DESC incident_count
GROUP BY animal_type
LIMIT 5;

ERROR: rapi_prepare: Failed to extract statements:
Parser Error: syntax error at or near "("

LINE 1: SELECT animal_type COUNT(*) AS incident_count
                                ^

SQL error messages can be very hard to decode, but AI is good at figuring out the issue. Try pressing the 'Fix & Explain' button to get an AI to help you - or ask it via the AI button on the right.
<br>
### Get the code to run with AI assistance

<br> AI is also good at explaining unfamiliar SQL code.

### Ask the AI to add an explanation to the bottom of this code

Answer: <br>
Add an explanation to the bottom of this code

In [23]:
SELECT animal_type, COUNT(*) AS incident_count
FROM 'animal_rescue_incidents.csv'
GROUP BY animal_type
ORDER BY incident_count DESC
LIMIT 5;

-- This query selects the top 5 animal types with the highest number of rescue incidents.
-- It groups the data by animal_type, counts the number of incidents for each type,
-- orders the results in descending order of incident count, and limits the output to 5 rows.

animal_type incident_count
1 cat         6884          
2 bird        2531          
3 dog         1756          
4 fox          762          
5 NA           265

## 7. Final Challenge
Now it's time to bring it all together.
### Ask the AI to show you the top 5 most rescued animals in Westminster.
> 🔍 Troubleshooting: If your code doesn't work first try, have a look at the table (the schema is above, and we saw some of the rows in Question 2) and check:
> * The table name is correct
> * The field names such as 'animal_type' are referenced correctly
> * The `WHERE` clause uses the correct casing (so you're not searching for 'westminster' instead of 'Westminster')

Answer: <br>
Using the dataset 'animal_rescue_incidents.csv', show me the top 5 most rescued animal_type instances where area = 'Westminster'.

In [26]:
SELECT animal_type, COUNT(*) AS rescue_count
FROM 'animal_rescue_incidents.csv'
WHERE area = 'Westminster'
GROUP BY animal_type
ORDER BY rescue_count DESC
LIMIT 5;

animal_type rescue_count
1 cat         215         
2 bird        119         
3 dog          33         
4 fox          13         
5 NA            7

## You've done it!
Congratulations on completing the exercises in this Webinar. You're well on your way to being a `SQL ROCKSTAR`!

<br>**Remember the AI Workflow 🤖**
<br>When solving a question, follow the following steps:
1. Think about the data you have and what you want to achieve.
2. Write a natural language prompt.
3. Understand the code generated by the AI.
4. Check the AI’s code for errors or hallucinations.
5. Run the code and verify the results.
6. Iterate.

Happy coding!